File: takkernel/engine/src/main/java/com/atakmap/map/layer/feature/cursor/BruteForceLimitOffsetFeatureCursor.java
package com.atakmap.map.layer.feature.cursor;

import com.atakmap.database.RowIteratorWrapper;
import com.atakmap.map.layer.feature.AttributeSet;
import com.atakmap.map.layer.feature.Feature;
import com.atakmap.map.layer.feature.FeatureCursor;

public class BruteForceLimitOffsetFeatureCursor extends RowIteratorWrapper implements FeatureCursor {

    private final FeatureCursor filter;
    private final int offset;
    private final int limit;
    private int pos;
    
    public BruteForceLimitOffsetFeatureCursor(FeatureCursor filter, int offset, int limit) {
        super(filter);
        
        this.filter = filter;

        this.pos = 0;
        this.offset = offset;
        this.limit = (limit > 0) ? limit : Integer.MAX_VALUE;
    }

    @Override
    public boolean moveToNext() {
        // if we've exceeded the limit, return false
        if((this.pos-this.offset) >= this.limit)
            return false;

        // fast forward to the first offset record
        while(this.pos < (this.offset-1)) {
            if(super.moveToNext())
                this.pos++;
            else
                return false;
        }
        
        // input exhausted
        if(!super.moveToNext())
            return false;

        // update the position and make sure we haven't exceeded the limit
        this.pos++;
        return ((this.pos-this.offset) <= this.limit);
    }
    
    @Override
    public Object getRawGeometry() {
        return this.filter.getRawGeometry();
    }

    @Override
    public int getGeomCoding() {
        return this.filter.getGeomCoding();
    }

    @Override
    public String getName() {
        return this.filter.getName();
    }

    @Override
    public int getStyleCoding() {
        return this.filter.getStyleCoding();
    }

    @Override
    public Object getRawStyle() {
        return this.filter.getRawStyle();
    }

    @Override
    public AttributeSet getAttributes() {
        return this.filter.getAttributes();
    }

    @Override
    public Feature get() {
        return this.filter.get();
    }

    @Override
    public long getId() {
        return this.filter.getId();
    }

    @Override
    public long getVersion() {
        return this.filter.getVersion();
    }

    @Override
    public long getFsid() {
        return this.filter.getFsid();
    }
}

File: takkernel/engine/src/main/cpp/feature/BruteForceLimitOffsetFeatureCursor.cpp
#include "feature/BruteForceLimitOffsetFeatureCursor.h"

using namespace TAK::Engine::Feature;

using namespace TAK::Engine::Util;

BruteForceLimitOffsetFeatureCursor::BruteForceLimitOffsetFeatureCursor(FeatureCursorPtr &&filter_, std::size_t limit_, std::size_t offset_) NOTHROWS :
    filter(std::move(filter_)),
    limit((limit_ > 0) ? limit_ : std::numeric_limits<std::size_t>::max()),
    offset(offset_),
    pos(0)
{}

TAKErr BruteForceLimitOffsetFeatureCursor::getId(int64_t *value) NOTHROWS
{
    return filter->getId(value);
}

TAKErr BruteForceLimitOffsetFeatureCursor::getFeatureSetId(int64_t *value) NOTHROWS
{
    return filter->getFeatureSetId(value);
}

TAKErr BruteForceLimitOffsetFeatureCursor::getVersion(int64_t *value) NOTHROWS
{
    return filter->getVersion(value);
}

TAKErr BruteForceLimitOffsetFeatureCursor::getRawGeometry(FeatureDefinition2::RawData *value) NOTHROWS
{
    return filter->getRawGeometry(value);
}

FeatureDefinition2::GeometryEncoding BruteForceLimitOffsetFeatureCursor::getGeomCoding() NOTHROWS
{
    return filter->getGeomCoding();
}

AltitudeMode BruteForceLimitOffsetFeatureCursor::getAltitudeMode() NOTHROWS 
{
    return filter->getAltitudeMode();
}

double BruteForceLimitOffsetFeatureCursor::getExtrude() NOTHROWS 
{
    return filter->getExtrude();
}

TAKErr BruteForceLimitOffsetFeatureCursor::getName(const char **value) NOTHROWS
{
    return filter->getName(value);
}

FeatureDefinition2::StyleEncoding BruteForceLimitOffsetFeatureCursor::getStyleCoding() NOTHROWS
{
    return filter->getStyleCoding();
}

TAKErr BruteForceLimitOffsetFeatureCursor::getRawStyle(FeatureDefinition2::RawData *value) NOTHROWS
{
    return filter->getRawStyle(value);
}

TAKErr BruteForceLimitOffsetFeatureCursor::getAttributes(const atakmap::util::AttributeSet **value) NOTHROWS
{
    return filter->getAttributes(value);
}

TAKErr BruteForceLimitOffsetFeatureCursor::get(const Feature2 **feature) NOTHROWS
{
    return filter->get(feature);
}

TAKErr BruteForceLimitOffsetFeatureCursor::moveToNext() NOTHROWS
{
    TAKErr code;

    // if we've exceeded the limit, return false
    size_t limitCompare = this->pos < this->offset ? this->offset : this->pos;
    if ((limitCompare - this->offset) >= this->limit)
        return TE_Done;

    code = TE_Ok;

    // fast forward to the first offset record
    while (this->pos < (this->offset /*- 1*/)) {
        code = filter->moveToNext();
        TE_CHECKBREAK_CODE(code);

        this->pos++;
    }
    TE_CHECKRETURN_CODE(code);

    // input exhausted
    code = filter->moveToNext();
    TE_CHECKRETURN_CODE(code);

    // XXX - looks broken

    // update the position and make sure we haven't exceeded the limit
    this->pos++;
    if ((this->pos - this->offset) <= this->limit)
        return TE_Ok;
    else
        return TE_Done;
}
